### PR TITLE
Blog - External Secret Operators (ESO) with HashiCorp Vault

### PR DESCRIPTION
**Pull request by msali123**
Done with all checks.

### CODE REVIEW

**Comment by msali123** [comment]
[Preview Link](https://642b767dab8bc42cee3e3da5--ephemeral-narwhal-3ac3b2.netlify.app/blog/eso-with-hashicorp-vault/)

@balapriyac

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+
+The External Secret Operator works by creating three custom resource definitions (CRDs); the *ClusterSecretStore*, the *SecretStore*, and the *ExternalSecret* CRD.
+
+The ClusterSecretStore CRD is used to define the connection details to the external secret store, such as the endpoint, authentication details, and other configuration settings. The SecretStore CRD is just like the ClusterSecretStore resource but it is namespaced. The ExternalSecret CRD is used to create and manage the Kubernetes secrets based on the configuration defined in either the ClusterSecretStore CRD or the SecretStore CRD.
+
+The External Secret Operator watches for changes to *ExternalSecret* resources and dynamically creates Kubernetes secrets as needed, populating them with the retrieved secret data. This way, the secrets remain secure and can be rotated regularly with no disruption to the applications using them.
+
+<div class="wide">
+![how eso works with vault]({{site.images}}{{page.slug}}/7Qinp9m.png)
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:44]:
```suggestion
![How ESO Works with Vault]({{site.images}}{{page.slug}}/7Qinp9m.png)
```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+<div class="wide">
+![Outputting vault version]({{site.images}}{{page.slug}}/bPU4iYh.png)
+</div>
+
+### Spinning Up a Vault Server
+
+Since you now have vault installed, the next step is to spin up a vault server and create some secrets in it. This step is important as the vault server serves as the main component of all our vault operations (creating secrets, deleting secrets, etc)
+
+Create a `[hcl` file](<https://hub.packtpub.com/what-is-hcl-hashicorp-configuration-language-how-does-it-relate-to-terraform-and-why-is-it-growing-in-popularity/>) and paste in the following configuration settings, you can name this file what you want, this tutorial uses `vault-config.hcl`.
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:111]:
```suggestion

Create a [`hcl` file](https://hub.packtpub.com/what-is-hcl-hashicorp-configuration-language-how-does-it-relate-to-terraform-and-why-is-it-growing-in-popularity/) and paste in the following configuration settings, you can name this file what you want, this tutorial uses `vault-config.hcl`.

```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+~~~
+
+If you have the below output then the file exists and is populated with the configuration settings:
+
+<div class="wide">
+![Confirming vault config file]({{site.images}}{{page.slug}}/UIDOt1x.png)
+</div>
+
+Create the `vault` directory that vault will use to store the data when running in server mode with the `[raft` storage backend](<https://developer.hashicorp.com/vault/docs/configuration/storage/raft>):
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:144]:
```suggestion

Create the `vault` directory that vault will use to store the data when running in server mode with the [`raft` storage backend](https://developer.hashicorp.com/vault/docs/configuration/storage/raft):

```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+</div>
+
+## Fetching Secrets from Hashicorp Vault With Eso
+
+Now that you have the Vault server up and have added some secret to it, you are now ready to use the External Secret Operator (ESO) to fetch and use that secret from your HashiCorp vault server.
+
+### Installing the External Secret Operator via Helm
+
+To use the External Secret operator, you need to first install it in your Kubernetes cluster. We will install it via [Helm]((<https://earthly.dev/blog/helm-k8s-mngr/>). You can check for other means of [installation guide](https://external-secrets.io/v0.7.2/introduction/getting-started/).
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:274]:
```suggestion

To use the External Secret operator, you need to first install it in your Kubernetes cluster. We will install it via [Helm](https://earthly.dev/blog/helm-k8s-mngr/). You can check for other means of [installation guide](https://external-secrets.io/v0.7.2/introduction/getting-started/).

```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+This command installs the *external-secrets* package using the Helm package manager. The package is sourced from the *external-secrets* repository and the installation process creates a new namespace named *external-secrets* and sets the *installCRDs* option to *true* so the Custom Resource Definitions for External Secrets Operator are installed with it (the *ClusterSecretStore*, *SecretStore* and *ExternalSecret* CRDs).
+
+~~~{.bash caption=">_"}
+helm install external-secrets \
+    external-secrets/external-secrets \
+    -n external-secrets \
+    --create-namespace \
+    --set installCRDs=true
+~~~
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:309]:
```suggestion
This command installs the *external-secrets* package using the Helm package manager. The package is sourced from the *external-secrets* repository and the installation process creates a new namespace named *external-secrets* and sets the *installCRDs* option to *true* so the Custom Resource Definitions for External Secrets Operator are installed with it (the *ClusterSecretStore*, *SecretStore* and *ExternalSecret* CRDs).
```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+<div class="wide">
+![Updating external-secret operator helm chart]({{site.images}}{{page.slug}}/2HmKRzC.png)
+</div>
+
+Now install the External Secret Operator using the following command:
+
+This command installs the *external-secrets* package using the Helm package manager. The package is sourced from the *external-secrets* repository and the installation process creates a new namespace named *external-secrets* and sets the *installCRDs* option to *true* so the Custom Resource Definitions for External Secrets Operator are installed with it (the *ClusterSecretStore*, *SecretStore* and *ExternalSecret* CRDs).
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:301]:
```suggestion



```

File: blog/_posts/2023-05-23-eso-with-hashicorp-vault.md [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+        property: POSTGRES_PASSWORD
+~~~
+
+This code defines a Custom Resource Definition (CRD) of type **`ExternalSecret`**
+ in the cluster.
+
+- The **`ExternalSecret`** resource specifies the interval at which to refresh the target secret, the name of the **`ClusterSecretStore`** that contains the secret, the target secret to be updated and the properties in the remote secret that will be mapped to the target secret.
+- The `**target**` field specifies the name of the target Kubernetes secret that the external secret ****data will be written to and the policy for creating it if it does not already exist.
+- The **`data`** field maps the properties in the remote secret to the target secret.
+
```
Comment (balapriyac) [inline blog/_posts/2023-05-23-eso-with-hashicorp-vault.md:445]:
```suggestion

- The **`ExternalSecret`** resource specifies the interval at which to refresh the target secret, the name of the **`ClusterSecretStore`** that contains the secret, the target secret to be updated and the properties in the remote secret that will be mapped to the target secret.
- The **`target`** field specifies the name of the target Kubernetes secret that the external secret data will be written to and the policy for creating it if it does not already exist.
- The **`data`** field maps the properties in the remote secret to the target secret.

```

**Review by balapriyac — CHANGES_REQUESTED** [review]
Thanks, Sameer.

**Comment by msali123** [comment]
@balapriyac 
[Done](https://642d5c5e7b8e351bfdf7064f--ephemeral-narwhal-3ac3b2.netlify.app/blog/eso-with-hashicorp-vault/)

**Comment by balapriyac** [comment]
@jalletto this is ready to merge.

**Review by jalletto — APPROVED** [review]

**Pull request merged**